File: makelinks.py
#!/usr/bin/python

import os
import sys
import zipfile

tigerpath = os.path.join(
	os.path.dirname(
		os.path.abspath(__file__)),
	'tiger')
if tigerpath not in sys.path:
	sys.path.append(tigerpath)

from record1 import record1

class linker(object):
	def __init__(self):
		self.they = {}
		self.halves = {}
		self.verbose = False
	
	def put(self, a, b):
		if a < b:
			self.they[(a,b)] = 1
		else:
			self.they[(b,a)] = 1
	
	def half(self, ubid, tlid):
		x = self.halves.get(tlid)
		if x is not None:
			if x == ubid:
				if self.verbose:
					sys.stderr.write('weird split, tlid=%09d ubid=%013u\n' % (tlid, ubid))
				return
			self.put(ubid,x)
			del self.halves[tlid]
		else:
			self.halves[tlid] = ubid
	
	def process(self, cur):
		"""cur is a record1"""
		numrecs = len(cur.raw) / record1.fieldwidth
		for i in xrange(0, numrecs):
			stl = cur.STATEL(i)
			countyl = cur.COUNTYL(i)
			tractl = cur.TRACTL(i)
			blockl = cur.BLOCKL(i)
			if (stl[0] == ' ') or (countyl[0] == ' ') or (tractl[0] == ' ') or (blockl[0] == ' '):
				ubidl = None
			else:
				ubidl = long(countyl + tractl + blockl)
			rst = cur.STATER(i)
			countyr = cur.COUNTYR(i)
			tractr = cur.TRACTR(i)
			blockr = cur.BLOCKR(i)
			if (rst[0] == ' ') or (countyr[0] == ' ') or (tractr[0] == ' ') or (blockr[0] == ' '):
				ubidr = None
			else:
				ubidr = long(countyr + tractr + blockr)
			if ubidl is None:
				if ubidr is None:
					sys.stderr.write('error, %s:%d has neither left nor reight\n' % ('', i+1))
				else:
					self.half(ubidr, int(cur.TLID(i)))
			else:
				if ubidr is None:
					self.half(ubidl, int(cur.TLID(i)))
				elif ubidr == ubidl:
					pass
					#sys.stderr.write('dup ubid %013u\n' % ubidl)
				else:
					self.put(ubidl, ubidr)
	
	def processZipFilename(self, fname):
		zf = zipfile.ZipFile(fname, 'r')
		for x in zf.namelist():
			if x.lower().endswith('.rt1'):
				if self.verbose:
					sys.stderr.write('reading %s(%s)\n' % (fname, x))
				raw = zf.read(x)
				cur = record1(raw)
				self.process(cur)
	
	def processFilename(self, fname):
		fi = open(fname, 'r')
		raw = fi.read()
		cur = record1(raw)
		self.process(cur)
	
	def writeText(self, out):
		for x in self.they.iterkeys():
			(a, b) = x
			out.write('%013u%013u\n' % (a, b))

def main(argv):
	ta = argv[1:]
	out = sys.stdout
	sourcelist = []
	while ta:
		arg = ta.pop(0)
		if arg == '-o':
			out = open(ta.pop(0), 'wb')
		else:
			sourcelist.append(arg)
	l = linker()
	l.verbose = True
	for s in sourcelist:
		if s.lower().endswith('.zip'):
			l.processZipFilename(s)
		elif s.lower().endswith('.rt1'):
			l.processFilename(s)
	l.writeText(out)
	out.close()

if __name__ == '__main__':
	main(sys.argv)

File: tiger/makeLinks.cpp
#include "record1.h"
#include "mmaped.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <ext/hash_set>

class blocklink {
public:
	uint64_t ubida, ubidb;
	blocklink( uint64_t a, uint64_t b ) {
		if ( a < b ) {
			ubida = a;
			ubidb = b;
		} else {
			ubida = b;
			ubidb = a;
		}
	}
};
class halflink {
public:
	uint64_t ubid;
	uint32_t tlid;
	halflink( uint64_t u, uint32_t t ) : ubid( u ) , tlid( t ) {}
};

// C++ is fugly
namespace __gnu_cxx {
	
template<> struct hash<blocklink> {
	size_t operator()( const blocklink& x ) const {
		if ( sizeof(uint64_t) > sizeof(size_t) ) {
			return (x.ubida && 0xffffffff) + (x.ubida >> 32) + (x.ubidb & 0xffffffff) + (x.ubidb >> 32);
		} else {
			return x.ubida + x.ubidb;
		}
	}
};

template<> struct hash<halflink> {
	size_t operator()( const halflink& x ) const {
		return x.tlid;
	}
};

}

namespace std {
	template<> struct equal_to<blocklink> {
		bool operator()( const blocklink& a, const blocklink& b ) const {
			return a.ubida == b.ubida && a.ubidb == b.ubidb;
		}
	};
	template<> struct equal_to<halflink> {
		bool operator()( const halflink& a, const halflink& b ) const {
			return a.tlid == b.tlid;
		}
	};
}

using namespace __gnu_cxx;

//template class hash_set<blocklink>;

hash_set<blocklink> they;

hash_set<halflink> halves;

class source {
public:
	int len;

	const char* filename;
	mmaped m;
	record1 r;
	
	source* next;
	source* prev;
	
	source( const char* f, source* n ) : len( 0 ), filename( f ), next( n ) {
		if ( next != NULL ) {
			prev = next->prev;
			next->prev = this;
			if ( prev != NULL ) {
				prev->next = this;
			}
		} else {
			prev = NULL;
		}
	}
};

#define memcpy( dst, src, len ) my_inline_memcpy( dst, src, len )
#define uint8_t unsigned char
inline void my_inline_memcpy( void* dst, const void* src, size_t len ) {
	uint8_t* du = (uint8_t*)dst;
	const uint8_t* su = (const uint8_t*)src;
	while ( len > 0 ) {
		*du = *su;
		du++;
		su++;
		len--;
	}
}

inline uint64_t ubid( unsigned long county, unsigned long tract, unsigned long block ) {
	uint64_t toret = county;
	assert( county <= 999 );
	assert( tract <= 999999 );
	assert( block <= 9999 );
	toret *= 1000000;
	toret += tract;
	toret *= 10000;
	toret += block;
	return toret;
}

void trypair( uint64_t a, uint64_t b ) {
	blocklink tbl( a, b );
	if ( they.count( tbl ) == 0 ) {
		they.insert( tbl );
	}
}

void tryhalf( uint64_t ubidr, uint32_t tlid ) {
	halflink nh( ubidr, tlid );
	hash_set<halflink>::iterator hi = halves.find( nh );
	if ( hi != halves.end() ) {
		halflink oh = *hi;
//		printf("found old half %u\n", oh.tlid );
		halves.erase( nh );
		if ( ubidr == oh.ubid ) {
			fprintf(stderr,"weird split, tlid=%u ubid=%013llu\n", tlid, ubidr );
		} else {
			trypair( ubidr, oh.ubid );
		}
	} else {
		halves.insert( nh );
	}
}

int main( int argc, const char** argv ) {
	source* root = NULL;
	const char* outname = NULL;
	FILE* out = NULL;

	for ( int i = 1; i < argc; i++ ) {
		if ( !strcmp( argv[i], "-o" ) ) {
			i++;
			outname = argv[i];
			out = fopen( outname, "wb" );
			if ( out == NULL ) {
				perror("fopen");
				exit(1);
			}
		} else {
			root = new source( argv[i], root );
		}
	}
	if ( out == NULL ) {
		printf("no output specified\n");
		exit(1);
		//out = stdout;
	}
	
	source* cur;
	cur = root;
	long recordsin = 0;
	long recordsout = 0;
	while ( cur != NULL ) {
		int err;
                fprintf(stderr,"%s: starting\n", cur->filename);
		err = cur->m.open( cur->filename );
		if ( err != 0 ) {
			perror("mmaped.open");
			exit(1);
		}
		cur->len = cur->m.sb.st_size / record1::size;
		cur->r.base = (uint8_t*)(cur->m.data);
		recordsin += cur->len;
		
		for ( int i = 0; i < cur->len; i++ ) {
			uint32_t tlid;
			uint64_t ubidl, ubidr;
			int nol, nor;
			tlid = cur->r.TLID_longValue(i);
			if ( cur->r.STATEL(i)[0] == ' ' || cur->r.COUNTYL(i)[0] == ' ' ||
				 cur->r.TRACTL(i)[0] == ' ' || cur->r.BLOCKL(i)[0] == ' ' ) {
				// left side empty
				nol = 1;
				ubidl = 0;
			} else {
				nol = 0;
				ubidl = ubid( cur->r.COUNTYL_longValue(i), cur->r.TRACTL_longValue(i), cur->r.BLOCKL_longValue(i) );
			}
			if ( cur->r.STATER(i)[0] == ' ' || cur->r.COUNTYR(i)[0] == ' ' ||
				 cur->r.TRACTR(i)[0] == ' ' || cur->r.BLOCKR(i)[0] == ' ' ) {
				// right side empty
				nor = 1;
				ubidr = 0;
			} else {
				nor = 0;
				ubidr = ubid( cur->r.COUNTYR_longValue(i), cur->r.TRACTR_longValue(i), cur->r.BLOCKR_longValue(i) );
			}
			if ( nol ) {
				if ( nor ) {
					printf("error, %s:%d has neither left nor right\n", cur->filename, i );
				} else {
					tryhalf( ubidr, tlid );
				}
			} else {
				if ( nor ) {
					tryhalf( ubidl, tlid );
				} else if ( ubidl != ubidr ) {
					trypair( ubidl, ubidr );
				}
			}
		}
		
		cur->m.close();
		cur = cur->next;
	}
	{
		hash_set<blocklink>::iterator bi = they.begin();
		hash_set<blocklink>::iterator end = they.end();
		while ( bi != end ) {
			fprintf(out,"%013llu%013llu\n",bi->ubida,bi->ubidb);
			bi++;
		}
		fflush( out );
	}
	recordsout = they.size();
	printf("%ld records in\n%ld links out\n%lu unpaired halves\n", recordsin, recordsout, (long unsigned int)(halves.size()) );
	fclose( out );
}
